### PR TITLE
Add type hints to `tests/rest/client`

### PR DESCRIPTION
**Pull request by dklimpel**
Wait for: #12084

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Thanks again. I'm generally happy assuming mypy is too, but have one style suggestion.

I'd be interested to how precise mypy's coverage of our test suite is. [mypy has some reporting options](https://matrix.org/blog/2021/12/17/type-coverage-for-sydent-evaluation#quantifying-coverage) which will produce summary statistics for this. Might be one to look into another time though; not suggesting we do this now.

File: tests/rest/client/test_register.py [hunk 1]
```diff
@@ -454,8 +460,8 @@ def test_POST_registration_token_session_expiry(self):
         )
 
         # Do 2 requests without auth to get two session IDs
-        params1 = {"username": "bert", "password": "monkey"}
-        params2 = {"username": "ernie", "password": "monkey"}
+        params1: JsonDict = {"username": "bert", "password": "monkey"}
+        params2: JsonDict = {"username": "ernie", "password": "monkey"}
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply tests/rest/client/test_register.py:464]:
```
tests/rest/client/test_register.py:471: error: Incompatible types in assignment (expression has type "Dict[str, Any]", target has type "str")  [assignment]
tests/rest/client/test_register.py:486: error: Unsupported target for indexed assignment ("str")  [index]
```

[the remark this answers is not in the document]
Comment (dklimpel) [reply tests/rest/client/test_register.py:464]: There is an initialization with
```python
params1 = {"username": "bert", "password": "monkey"}
```

And a modification with:
```python
        params1["auth"] = {
            "type": LoginType.REGISTRATION_TOKEN,
            "token": token,
            "session": session1,
        }
```

[the remark this answers is not in the document]
Comment (DMRobertson) [reply tests/rest/client/test_register.py:464]: That's fair!

**Comment by dklimpel** [comment]
> mypy isn't happy in `test_groups.py:40`. It's pasing `tok=None` but `tok: str`. I think the annotation should be `tok: Optional[str]` maybe? Either that or pass `tok=''`

It needs PR #12084

**Comment by dklimpel** [comment]
It passes now.

**Review by DMRobertson — APPROVED** [review]
Thank you!

**Pull request merged**